### PR TITLE
Add nlprob to ODEFunction

### PR DESCRIPTION
**Pull request by oscardssmith**

### CODE REVIEW

File: src/scimlfunctions.jl [hunk 1]
```diff
@@ -289,6 +289,9 @@ the usage of `f`. These include:
   based on the sparsity pattern. Defaults to `nothing`, which means a color vector will be
   internally computed on demand when required. The cost of this operation is highly dependent
   on the sparsity pattern.
+- `nlfunc`: a `NonlinearFunction`
+- `nl_state_compres`: maps u->nlfunc_u
+- `nl_state_decompres`: maps nlfunc_u->u
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:294]:
```suggestion
- `nl_state_compress`: maps u->nlfunc_u
- `nl_state_decompress`: maps nlfunc_u->u
```

**Comment by oscardssmith** [comment]
The downstream CI failures don't make sense to me. It looks like the downstream is only seeing some of the changes from https://github.com/SciML/SciMLBase.jl/pull/816, but I don't really understand it. Closing and reopening to see if that makes CI happier.

**Comment by ChrisRackauckas** [comment]
Is this ready @oscardssmith

**Comment by oscardssmith** [comment]
no it needs a quick fix that I'll get to this morning. I'll ping when ready.

**Comment by oscardssmith** [comment]
OK I now think this should work.

**Comment by ChrisRackauckas** [comment]
Looks good to go but needs a real docstring 😅

File: src/scimlfunctions.jl [hunk 2]
```diff
@@ -289,6 +289,7 @@ the usage of `f`. These include:
   based on the sparsity pattern. Defaults to `nothing`, which means a color vector will be
   internally computed on demand when required. The cost of this operation is highly dependent
   on the sparsity pattern.
+- `nlfunc`: a `NonlinearFunction`
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/scimlfunctions.jl]: docs actually added

**Comment by oscardssmith** [comment]
changed to be a nlprob instead of nlfunc (to be closer to initializeprob).

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
It's still not sufficiently documented... but I'm not going to hold it up anymore. Come back to document it when we have a clearer sense of what it is.

**Comment by oscardssmith** [comment]
what else do you want for docs?

**Comment by ChrisRackauckas** [comment]
Define mathematically what that problem is, what its parameters are, etc.

**Pull request merged**